tests/Agent/UnitTests/Core.UnitTest/Wrapper/AgentWrapperApi/Builders/TransactionTests.cs

```diff
@@ -296,5 +296,54 @@ public void AddUserAttribute_LastInWins()
 
             Assert.AreEqual(result, valueB);
         }
+
+        #region User Tracking Tests
+        [Test]
+        public void UserTracking_NoEndUserIdAttributeIfNotSet()
+        {
+            var immutableTransactionMetadata = _transaction.TransactionMetadata.ConvertToImmutableMetadata();
+
+            var userAttributes = immutableTransactionMetadata.UserAndRequestAttributes.ToDictionary();
+            var hasUserIdAttribute = userAttributes.TryGetValue(_attribDefs.EndUserId.Name, out var userIdValue);
+            Assert.False(hasUserIdAttribute);
+        }
+
+        [Test]
+        [TestCase("CustomUserId", true)]
+        [TestCase("", false)]
+        [TestCase(" ", false)]
+        [TestCase(null, false)]
+
+        public void UserTracking_ShouldAddEndUserIdAttributeWhenNotNullOrWhitespace(string expectedUserId, bool endUserAttributeShouldExist)
+        {
+            _transaction.SetUserId(expectedUserId);
+
+            var immutableTransactionMetadata = _transaction.TransactionMetadata.ConvertToImmutableMetadata();
+            var userAttributes = immutableTransactionMetadata.UserAndRequestAttributes.ToDictionary();
+            var hasUserIdAttribute = userAttributes.TryGetValue(_attribDefs.EndUserId.Name, out var userIdValue);
+
+            Assert.AreEqual(endUserAttributeShouldExist, hasUserIdAttribute);
+            if (endUserAttributeShouldExist)
+            {
+                Assert.AreEqual(expectedUserId, userIdValue);
+            }
+        }
+
+        [Test]
```

Comment:
Might be overkill, but should we test the case where we set the user ID to null after it previously had a value?

Reply:
Having a test like that might serve as additional documentation of that behavior.

Reply:
So I added a test for this and find that the previously-set user id is still present in the attributes after being set to null (since the code short-circuits when the value is null). 

I wonder if we should actually ensure the attribute gets removed when setting to null or whitespace?

Reply:
I thought that was behavior that we chose to remove in a previous PR. Either way, it's a good idea to confirm the intention.

Reply:
After thinking about it for a minute, I don't see any valid use case where the user would want to set a user id and then clear it later during the same transaction. So for now, we'll leave things alone.